File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9K.scala
package TextbookTrack.Chapter09

object BA9K {
  import scala.annotation.tailrec

  final case class CharacterIndex(char: Char, number: Int)

  private def createIndexedColumn(string: List[Char]): Vector[CharacterIndex] = {
    @tailrec
    def loop(counts: Map[Char, Int],
             charIndex: List[CharacterIndex],
             xs: List[Char]): Vector[CharacterIndex] = xs match {
      case Nil => charIndex.reverseIterator.toVector
      case x :: xss =>
        val count: Int = counts.getOrElse(x, 0)
        loop(counts.updated(x, count + 1), CharacterIndex(x, count) :: charIndex, xss)
    }

    loop(Map(), Nil, string)
  }

  def calcLastToFirstMapping(transformedText: List[Char], ix: Int): Int = {
    val lastColumn: Vector[CharacterIndex] = createIndexedColumn(transformedText)
    val firstColumn: Map[CharacterIndex, Int] = createIndexedColumn(transformedText.sorted).zipWithIndex.toMap
    firstColumn(lastColumn(ix))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val transformedText: List[Char] = reader.next().toList
    val ix: Int = reader.next().toInt
    val result: Int = calcLastToFirstMapping(transformedText, ix)
    println(result)
  }
}

File: Python/textbook_track/chapter09/ba9k.py
# Generate the Last-to-First Mapping of a String
import sys

from collections import defaultdict


def create_indexed_column(string):
    counts = defaultdict(int)
    char_index = []
    for letter in string:
        count = counts[letter]
        char_index.append((letter, count))
        counts[letter] += 1
    return char_index


def calc_first_column_with_index(transformed_string):
    first_column = create_indexed_column(sorted(transformed_string))
    return {char_index: ix for ix, char_index in enumerate(first_column)}


def calc_last_to_first_mapping(transformed_string, ix):
    last_column = create_indexed_column(transformed_string)
    first_column = calc_first_column_with_index(transformed_string)
    return first_column[last_column[ix]]


def main():
    reader = sys.stdin
    transformed_text = next(reader).rstrip()
    ix = int(next(reader))
    result = calc_last_to_first_mapping(transformed_text, ix)
    print(result)


if __name__ == "__main__":
    main()
